Add monitoring and logs connection examples (#909)
memgraph/documentation: pages/database-management/monitoring.mdx

# Monitoring

Monitoring of applications and databases is essential to be able to track the

> load, resources, logs, transactions, latency etc. 

Reviewer: Avoid using "etc". Give the correct, exhaustive list of things the user must consider.

Memgraph currently supports:

> - Logs tracking via WebSocket server 

Reviewer 2: Maybe `real-time logs`?

Author: Agreed. 👍

> - Monitoring via HTTP server (Enterprise Edition)

Reviewer 3: Add links to sections (anchors)

Reviewer 2: Does it make sense to say "Telemetry" or "Metrics" instead of monitoring? Because monitoring is a big term that can have a lot in it, e.g. logs are in it, but logs are not on HTTP.

> Memgraph currently supports:
> 
> - Logs tracking via WebSocket server 
> - Monitoring via HTTP server (Enterprise Edition)
> 
> Log tracking is helpful for debugging purposes and monitoring the database
> operations.  
>  
> In the Enterprise Edition, on top of log tracking,  Memgraph allows tracking
> information about transactions, query latencies, snapshot recovery latencies,
> triggers, bolt messages, indexes, etc., using an HTTP server.

Reviewer 4: Just a suggestion for a bit clearer structure:
```suggestion
Memgraph currently supports:
1. **Log tracking via WebSocket server**: Log tracking is helpful for debugging purposes and monitoring the database operations. 
2. **Monitoring via HTTP server (Enterprise Edition)**: In the Enterprise edition, besides log tracking, Memgraph allows tracking information about transactions, query latencies, snapshot recovery latencies, triggers, bolt messages, indexes, and more, using an HTTP server.
```

Reviewer 3: I agree, wanted to suggest the same

Reviewer 2: Agree!

Reviewer: Please avoid using terms such as "etc" and "and more". Prefer being thorough in listing everything the user might need or consider.

> ## Logs tracking via WebSocket server 

Reviewer 4: For the consistency:
```suggestion
## Log tracking via WebSocket server 
```

Author: Going with the @matea16 recommendation.

Connect to Memgraph's logging server via WebSocket to forward logs to all the
connected clients.

### Log messages

Each log that is written to the [log file](logs.mdx#logs) is forwarded to the
connected clients in the following format:

```json
{
  event: "log",
  level: "trace"|"debug"|"info"|"warning"|"error"|"critical",
  message: "<log-message>"
}
```

### Connection

To connect to Memgraph's WebSocket server, use the following URL:

```plaintext
ws://host:port
```

The default host is `0.0.0.0`, but it can be changed using the
`--monitoring-address=` configuration flag.  

The default port is `7444`, but it can be changed using the `--monitoring-port`
configuration flag.

To connect to Memgraph's WebSocket server using the default configuration and
Python client, you can use the following code snippet:

> ```Python

Reviewer 3: ```suggestion
```python
```

code highlighting is not working, check if it works before commiting

Author: Yeah, my last Grammarly check ruined this and the indents.

import asyncio
import websockets

HOST = 'localhost'
PORT = '7444'

# The WebSocket URL
ws_url = f'ws://{HOST}:{PORT}'

async def read_logs():
    async with websockets.connect(ws_url) as websocket:
        print(f"Connected to Memgraph at {ws_url}")
        try:
            # Keep reading messages (logs) from the server
            while True:

>  log_message = await websocket.recv()

Reviewer 2: What happened with the formating of this line indent?

Author: Good catch; it looks like Grammarly or Docs formatted destroyed it.

                print(log_message)
        except websockets.exceptions.ConnectionClosed:
            print("Connection to Memgraph closed.")

asyncio.run(read_logs())
```

The code structure is similar in other languages, depending on the WebSocket
library used.

If you want to connect with a WebSocket Secure (WSS) connection. Set the
following configuration flags to enable an SSL connection: `--bolt-cert-file`

> and `--bolt-key-file`.

Reviewer 3: smth is not right here, fix the sentence

Author: Done.

These flags are also used to set up an SSL connection when connecting via Bolt
protocol, hence the name. Refer to the configuration page to learn how to

> [update the

Author: Removed.

configuration](/database-management/configuration#changing-configuration) and to
see the list of all available [configuration
flags](/database-management/configuration#list-of-configuration-flags).

### Authentication

> If authentication is not used (there are no users present in Memgraph), no

Reviewer: s/If/When
s/there are/due to
Remove parenthesis

```Python
import asyncio
import json
import websockets

HOST = "localhost"
PORT = "7444"
USERNAME = "memgraph"
PASSWORD = "memgraph"

ws_url = f"ws://{HOST}:{PORT}"

async def authenticate_and_read_logs(websocket):
    # Send authentication credentials
 credentials = json.dumps({"username": USERNAME, "password": PASSWORD})
    await websocket.send(credentials)

    # Wait for authentication response
 response = await websocket.recv()
 response_data = json.loads(response)
    if response_data.get("success"):
        print("Authentication successful!")
    else:
        print(f"Authentication failed: {response_data.get('message')}")
        return  # Stop if authentication fails

    # After successful authentication, start reading logs
    try:
        while True:
 log_message = await websocket.recv()
            print(log_message)
    except websockets.exceptions.ConnectionClosed:
        print("Connection to Memgraph closed.")

async def connect_and_authenticate():
    async with websockets.connect(ws_url) as websocket:
        print(f"Connected to Memgraph at {ws_url}")
        await authenticate_and_read_logs(websocket)

asyncio.run(connect_and_authenticate())
```

> #### Authorization (Enterprise)

Reviewer 3: is it h4 on purpose?

Author: Yeah, it is related to the WEB socket auth access.

Permission for connecting through WebSocket is controlled by the [`WEBSOCKET`
privilege](/database-management/authentication-and-authorization/role-based-access-control#privileges).

## Monitoring via HTTP server (Enterprise)

In the Enterprise Edition, Memgraph allows tracking information about
transactions, query latencies, snapshot recovery latencies, triggers, Bolt

> messages, indexes, streams, and many more metrics using an HTTP server.

Author: Done.

> To be able to retrieve data from the HTTP server, [enter a valid Memgraph

Reviewer 3: ```suggestion
To retrieve data from the HTTP server, [enter a valid Memgraph
```

Enterprise license key](/database-management/enabling-memgraph-enterprise).  

The default address and port for the metrics server is `0.0.0.0:9091`, and can
be configured using the `--metrics-address` and `--metrics-port` [configuration
flags](/database-management/configuration#list-of-configuration-flags).

To retrieve the metrics, send a GET request to the following URL:

```plaintext
http://host:port
```

Here is an example of how to retrieve metrics using Python:

> ```Python

Reviewer 3: ```suggestion
```python
```

import requests
import json

def fetch_memgraph_metrics():
 metrics_url = "http://0.0.0.0:9091/metrics"

    try:
 response = requests.get(metrics_url, timeout=5) 

        if response.status_code == 200:
            try:
 metrics_data = json.loads(response.text)

>  pretty_metrics = json.dumps(metrics_data, indent=4)

Reviewer 2: Indents are screwed.

 | average_degree | Gauge (double)   | Average number of relationships of a single node.           |
 | memory_usage   | Gauge (uint64_t) | Amount of RAM used reported by the OS (in bytes).           |
 | disk_usage     | Gauge (uint64_t) | Amount of disk space used by the [data directory](/fundamentals/data-durability) (in bytes). |
 | peak_memory_usage | Gauge (uint64_t) | Peak RAM memory usage in the system during the whole run (in bytes).  |

#### Index metrics

 | Name                       | Type    | Description                                            |
 | -------------------------- | ------- | ------------------------------------------------------ |
 | ActiveLabelIndices         | Counter | Number of active label indexes in the system.          |
 | ActiveLabelPropertyIndices | Counter | Number of active label property indexes in the system. |

>  | ActiveTextIndices          | Counter | Number of active text indexes in the system.           |

Reviewer 2: I know I am late to the party, but does anyone know why are these `Counter`? This should be `Gauge` because it can go up or down. Counter should always go up or have a reset to 0.

> A counter is a cumulative metric that represents a single [monotonically increasing counter](https://en.wikipedia.org/wiki/Monotonic_function) whose value can only increase or be reset to zero on restart.

        "UnwindOperator": 0
 },
    "Query": {
        "QueryExecutionLatency_us_50p": 0,
        "QueryExecutionLatency_us_90p": 0,
        "QueryExecutionLatency_us_99p": 0
 },
    "QueryType": {
        "ReadQuery": 0,
        "ReadWriteQuery": 0,
        "WriteQuery": 0
 },
    "Session": {
        "ActiveBoltSessions": 0,
        "ActiveSSLSessions": 0,
        "ActiveSessions": 0,
        "ActiveTCPSessions": 0,
        "ActiveWebSocketSessions": 0,
        "BoltMessages": 0
 },
    "Snapshot": {
        "SnapshotCreationLatency_us_50p": 4860,
        "SnapshotCreationLatency_us_90p": 4860,
        "SnapshotCreationLatency_us_99p": 4860,
        "SnapshotRecoveryLatency_us_50p": 628,
        "SnapshotRecoveryLatency_us_90p": 628,
        "SnapshotRecoveryLatency_us_99p": 628
 },
    "Stream": {
        "MessagesConsumed": 0,
        "StreamsCreated": 0
 },
    "Transaction": {
        "ActiveTransactions": 0,
        "CommitedTransactions": 0,
        "FailedPrepare": 0,
        "FailedPull": 0,
        "FailedQuery": 0,
        "RollbackedTransactions": 0,
        "SuccessfulQuery": 0
 },
    "Trigger": {
        "TriggersCreated": 0,
        "TriggersExecuted": 0
 }
}
```

> ## Prometeus exporter

Reviewer: Spelling error
